File: android/src/main/java/com/navigine/naviginesdk/NavigineCircleMapObject.java
package com.navigine.naviginesdk;

import android.content.Context;

import androidx.annotation.NonNull;

import com.navigine.idl.java.AnimationType;
import com.navigine.idl.java.CircleMapObject;
import com.navigine.idl.java.LocationPoint;

import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NavigineCircleMapObject implements MethodChannel.MethodCallHandler {
  private final int id;
  private final Context context;
  private final CircleMapObject mapObject;
  private final MethodChannel methodChannel;

  public NavigineCircleMapObject (
    int id,
    Context context,
    CircleMapObject mapObject,
    BinaryMessenger messenger
  ) {
    this.id = id;
    this.mapObject = mapObject;
    this.context = context;

    methodChannel = new MethodChannel(messenger, "navigine_sdk/navigine_circle_map_object_" + id);
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "setPosition":
        result.success(setPosition(call));
        break;
      case "setPositionAnimated":
        result.success(setPositionAnimated(call));
        break;
      case "setColor":
        result.success(setColor(call));
        break;
      case "setRadius":
        result.success(setRadius(call));
        break;
      case "getType":
        result.success(getType(call));
        break;
      case "setVisible":
        result.success(setVisible(call));
        break;
      case "setInteractive":
        result.success(setInteractive(call));
        break;
      case "setStyle":
        result.success(setStyle(call));
        break;
      case "setData":
        setData(call);
        result.success(null);
        break;
      case "getData":
        result.success(getData(call));
        break;
      case "setTitle":
        result.success(setTitle(call));
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  public CircleMapObject getMapObject() {
    return mapObject;
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setPosition(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return mapObject.setPosition(Utils.locationPointFromJson((Map<String, Object>) params.get("locationPoint")));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setPositionAnimated(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    LocationPoint locationPoint = Utils.locationPointFromJson((Map<String, Object>) params.get("locationPoint"));
    float duration = ((Double) params.get("duration")).floatValue();
    int type = ((Integer) params.get("type"));
    return mapObject.setPositionAnimated(locationPoint, duration, AnimationType.values()[type]);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setColor(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Double red = (Double) params.get("red");
    Double green = (Double) params.get("green");
    Double blue = (Double) params.get("blue");
    Double alpha = (Double) params.get("alpha");
    return mapObject.setColor(red.floatValue(), green.floatValue(), blue.floatValue(), alpha.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setRadius(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Double radius = (Double) params.get("radius");
    return mapObject.setRadius(radius.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int getType(MethodCall call) {
    return mapObject.getType().ordinal();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setVisible(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean visible = (Boolean) params.get("visible");
    return mapObject.setVisible(visible);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setInteractive(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean interactive = (Boolean) params.get("interactive");
    return mapObject.setInteractive(interactive);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setStyle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String style = (String) params.get("style");
    return mapObject.setStyle(style);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setData(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    byte[] data = (byte[]) params.get("style");
    mapObject.setData(data);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public byte[] getData(MethodCall call) {
    return mapObject.getData();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setTitle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String title = (String) params.get("title");
    return mapObject.setTitle(title);
  }
}

File: ios/Classes/NavigineCircleMapObject.swift
import Foundation
import Navigine

public class NavigineCircleMapObject: NSObject {
    private var id: Int32
    private var mapObject: NCCircleMapObject
    private let methodChannel: FlutterMethodChannel!
    private let pluginRegistrar: FlutterPluginRegistrar!

    public required init(
        id: Int32,
        mapObject: NCCircleMapObject,
        registrar: FlutterPluginRegistrar
    ) {
        self.id = id
        self.mapObject = mapObject

        methodChannel = FlutterMethodChannel(
            name: "navigine_sdk/navigine_circle_map_object_\(id)",
            binaryMessenger: registrar.messenger()
        )

        self.pluginRegistrar = registrar

        super.init()

        weak var weakSelf = self
        self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setPosition":
            result(setPosition(call))
        case "setPositionAnimated":
            result(setPositionAnimated(call))
        case "setColor":
            result(setColor(call))
        case "setRadius":
            result(setRadius(call))
        case "getType":
            result(getType(call))
        case "setVisible":
            result(setVisible(call))
        case "setInteractive":
            result(setInteractive(call))
        case "setStyle":
            result(setStyle(call))
        case "setData":
            setData(call)
            result(nil)
        case "getData":
            result(getData(call))
        case "setTitle":
            result(setTitle(call))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func getMapObject() -> NCCircleMapObject {
        return mapObject
    }

    public func setPosition(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let locationPoint = Utils.locationPointFromJson(params["locationPoint"] as! [String : Any])

        return mapObject.setPosition(locationPoint)
    }

    public func setPositionAnimated(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let locationPoint = Utils.locationPointFromJson(params["locationPoint"] as! [String : Any])
        let duration = (params["duration"] as! NSNumber)
        let type = params["type"] as! Int

        return mapObject.setPositionAnimated(locationPoint, duration: duration.floatValue, type: NCAnimationType.init(rawValue: type)!)
    }

    public func setColor(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let red = params["red"] as! NSNumber
        let green = params["green"] as! NSNumber
        let blue = params["blue"] as! NSNumber
        let alpha = params["alpha"] as! NSNumber

        return mapObject.setColor(red.floatValue, green: green.floatValue, blue: blue.floatValue, alpha: alpha.floatValue)
    }

    public func setRadius(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let radius = (params["radius"] as! NSNumber)

        return mapObject.setRadius(radius.floatValue)
    }

    public func getType(_ call: FlutterMethodCall) -> Int {
        return mapObject.getType().rawValue
    }

    public func setVisible(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let visible = (params["visible"] as! NSNumber)

        return mapObject.setVisible(visible.boolValue)
    }

    public func setInteractive(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let interactive = (params["interactive"] as! NSNumber)

        return mapObject.setInteractive(interactive.boolValue)
    }

    public func setStyle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let style = (params["style"] as! String)

        return mapObject.setStyle(style)
    }

    public func setData(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let data = (params["data"] as! FlutterStandardTypedData).data

        return mapObject.setData(data)
    }

    public func getData(_ call: FlutterMethodCall) -> Data {
        return mapObject.getData()
    }

    public func setTitle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let title = (params["title"] as! String)

        return mapObject.setTitle(title)
    }

    private func getCircleImage(_ image: [String: Any]) -> UIImage {
        let type = image["type"] as! String
        let defaultImage = UIImage()

        if (type == "fromAssetImage") {
          let assetName = self.pluginRegistrar.lookupKey(forAsset: image["assetName"] as! String)

          return UIImage(named: assetName) ?? defaultImage
        }

        if (type == "fromBytes") {
          let imageData = (image["rawImageData"] as! FlutterStandardTypedData).data

          return UIImage(data: imageData) ?? defaultImage
        }

        return defaultImage
      }
}
